articles/containerization-and-why-to-use-it.md

```diff
@@ -0,0 +1,58 @@
+* Your application dependencies differ in version or feature set between environments.
+* Developers have difficulty reproducing production issues on their local machines.
+* Application dependencies are installed manually, developers spend days setting up their own machine.
+
+This article will explain how containerization can help solve each of these issues.
+
+## Cumbersome deployments and dirty hosts
+
+Running scripts on a host seems OK in the beginning, but over time your host will become "dirty". Some versions of your script were only run in QA, and production has leftover artifacts from previous deployments. These leftovers confuse maintainers and reduce confidence in your team's ability to deploy software to production systems.
+
+With containerization, you immediately get rid of leftovers your software may cause. By **throwing away the old container** after deploying a new version you have already cleaned up after your application.
+
+A tangent about volumes/host mounts: Sometimes your app writes files outside of its container. In these cases the files still need to be cleaned up, but usually an app has a well-defined output directory that rarely changes.
+
+Volumes allow you to use containerization with databases quite nicely: The data is safe from crashes but up- and downgrading a DB is just a question of stopping the old container and booting a new one.
+
+## More hardware and the procurement process
+
+There is a conflict between needing more resources and a company's willingness to spend money. Usually, there is a long-winded procurement process for ordering hardware and getting it installed it in a rack. In today's time where hardware is cheap and humans are expensive paying some PaaS a few hundred bucks a month is cheaper in most cases.
+
+Any established platform will offer you support for Docker, and if you have containerized your application you'll be able to solve the scaling problem a lot easier. Usually, this involves uploading the image to the platform, creating a service and then defining scaling rules for said service. While the platform will not run on your local machine this does not prevent you from testing locally, as you can install Docker and are guaranteed the same runtime as on the platform.
+
+Protip: Most platforms offer some functionality for doing rolling updates, you can deregister your containers from the loadbalancer before killing them thus achieving deployments that are invisible to the customer. Now every day is a production deployment day!
+
+## Different environments, different dependencies and the highway to hell
+
+While differences between environments should be as small they tend to increase over time. Someone upgrades only the QA database for testing; someone applies emergency security fixes only to production because other environments are not critical. You find an increasing number of builds pass QA but fail in staging or production!
```

Comment:
"While differences between environments should be as small they tend to increase over time." hä?
->
"Difference between environments should be as small as possible, unfortunately they tend to increase over time."